standards/core/rln-contract.md

```diff
@@ -0,0 +1,209 @@
+3. further upgrades, if necessary, can be done by deploying a new contract and migrating the membership set.
+
+## Parameters
+
+We make the following design choices:
+- there are `3` membership tiers: low-, mid-, and high-tier;
+- prices are quoted in `USD`;
+- payment is accepted in `DAI`;
+- the price for a given tier is linearly proportional to its rate limit.
+
+For the initial deployment, we suggest the following values.
+
+| Parameter                                                 | Symbol     | Value   | Units                                                                                | Modifiable? | Comment                        |
+| --------------------------------------------------------- | ---------- | ------- | ------------------------------------------------------------------------------------ | ----------- | ------------------------------ |
+| Epoch length                                              | `epoch`    | `10`    | minutes                                                                              | Yes         |                                |
+| Maximum total rate limit of concurrent active memberships | `R_{max}`  | TBD     | messages per `epoch`                                                                 | Yes         | Can be replaced with `N_{max}` |
+| Maximum number of concurrent active memberships           | `N_{max}`  | `10000` | units                                                                                | Yes         | Can be replaced with `R_{max}` |
+| Rate limit for low-tier                                   | `R_{low}`  | `20`    | messages per `epoch`                                                                 | TBD         |                                |
+| Rate limit for mid-tier                                   | `R_{mid}`  | `200`   | messages per `epoch`                                                                 | TBD         |                                |
+| Rate limit for high-tier                                  | `R_{high}` | `600`   | messages per `epoch`                                                                 | TBD         |                                |
+| Unit price                                                | `p_u`      | TBD     | `USD` per `1` message per `epoch` for the duration of one membership expiration term | Yes         |                                |
+| Membership expiration term                                | `T`        | `90`    | days                                                                                 | Yes         |                                |
+| Grace period                                              | `G`        | `30`    | days                                                                                 | Yes         |                                |
+
+### Discussion on some parameter values
+
+#### Epoch length
+
+Epoch length is a global parameter set in the smart contract.
```

Comment:
I'd say the main benefit of incoding an epoch in the contract is that we can piggyback on the underlying ~~timechain~~ blockchain as a time signal. We need at least a rough consensus on what the current time is, otherwise (in a hypothetical future protocol version with slashing) innocent nodes may be punished because of a network split w.r.t. what the current epoch is.

Reply:
Sorry reopening that. I think there are 2 decisions here being discussed.

1. What epoch length do nodes use to verify message validity on gossipsub
2. What current time do nodes use to verify message validity on gossipsub.

I am assuming @alrevuelta was referring to (1). Currently, the epoch length is harcoded in Waku client code, specifically on cluster parameters.
The question is whether this is moved to the smart contract. Meaning that nodes should query/listen to smart contract to decide epoch length.

I thinks whether we do it in code or smart contract depends on 2 factors:

a. How do we expect a change of this value to happen for an existing cluster
b. Do we expect different clusters to use same RLN contract, and if so, with same epoch length value.

It seems to me that (a) becomes very hard if we keep the length in the code. As a different value leads to network split. Meaning reducing stability of the network during a migration.
Hence, I think it would make sense to have the value in (a), probably accompanied with a "start time". Meaning that nodes can be informed of the new value from the smart contract and from when this new value needs to be applied.

In terms of (b), who do we expect to start a new cluster with same smart contract? us/Status? or external project? In any case, then could deploy a smart contract that delegate all calls to original smart contract, except for the `getEpochLength` function and set their own.

Hence, it seems that setting epoch length in the smart contract is the most future proof option.

---

Regarding (2). Due to the granularity of block time (20s on mainnet? how long on L2s?), I am not convinced it would work for gossipsub. If that's something we want to do, we will need to study it properly first. So I'd suggest no for now, and we can book the work if we see it necessary.

Reply:
Regarding (1), I agree that

> setting epoch length in the smart contract is the most future proof option.

I'm not sure I understand this point:

> Due to the granularity of block time (20s on mainnet? how long on L2s?), I am not convinced it would work for gossipsub. If that's something we want to do, we will need to study it properly first.

What do you mean by "it"? How do nodes currently understand what the time is?

Moreover, it seems to me that there is a third question here:

3. How do nodes know what the current epoch is to verify message validity on gossipsub

Just knowing the epoch length and the current time is insufficient to determine what the current epoch is. The node would additionally need to know either of:

a. the global timestamp at which the first epoch began (and derive the current epoch locally); or
b. the current epoch as reported by other nodes (or by the smart contract? or by the blockchain itself via its block headers?).

AFAIU, each message (in particular, its proof) is committed to a particular epoch. My high-level concern here is that message validity depends on Relay nodes being (more-or-less?) in sync w.r.t. what the current epoch is. Lacking a full-fledged consensus, we cannot guarantee synchronization, unless we piggyback on the underlying blockchain (L1, L2, doesn't matter) that provides this functionality by design (at least in terms of block heights if not physical timestamps).

Does it make sense, or am I misunderstanding something?

Reply:
Now I finally understand how this works (h/t @rymnc):
1. we assume every node has (more or less) precise time signal;
1. every node derives the epoch number by dividing the current UNIX timestamp by epoch length ([code](https://github.com/waku-org/nwaku/blob/71d4a11647a905d16340c5ba4d95a6082e6cda0a/waku/waku_rln_relay/rln_relay.nim#L93-L97)).

In other words, RLN epoch numbers don't (and don't have to) start with zero or one.

For now I am leaving the spec as is w.r.t encoding epoch length in the contract as "the most future proof" option.